README.txt
README for Build PNG tools

build-png-tools-v###.zip

ALL THE FILES IN THIS ZIP:
changelog.txt				->		change log for each major version
ideas.txt					->		ideas for future versions
lic-gpl3.txt				->		License for my code
readme.txt					->		This file
PALETTE.DAT					->		test palette for reference (modified a bit, still uses similar color ranges though)
source\art2png.c			->		Source C file for art2png
source\png2art.c			->		Source C file for png2art
source\trythis.c			->		Experiment for working directories, not needed to be compiled.
palettes\duke3d_normal.act	->		Photoshop Raw Color Table for making PNGs with
palettes\duke3d_alt.act		->		Photoshop Raw Color Table for making PNGs with (slightly different, less saturated)

DESCRIPTION:

Observing the Duke 3D community, I wanted to help people edit Duke 3D's tiles in a way that would be quick yet result in classic ART tiles.

See, eDuke32 is great, don't get me wrong. You can create PNGs and use those as tiles using a bunch of tilefromtexture commands.

But if you're going to replace many graphics, like I do, it's better to go vanilla. One problem - no tools exist aside EditART and BastART, which are showing their age.

Sure, there's art2tga - which does it's job - but tga2art always doesn't work, due to it's admittedly messy way it handles tiles.

What I did was I improved these tools by rewriting them to handle PNGs using FreeImage, which works better IMO.

No point beating around the bush, my code isn't that well commented and at times is quite ugly.

There is no excuse, but if I had to pull one out of my behind it's the fact I mainly used ReBuild TGA-ART tools code as a base/reference, which wasn't well commented. Oh well.

Right now it only supports Duke 3D's tiles - which means increments of 256 per tilexxx.art. You can change this for Blood support or other build games.

I have since removed the pre-compiled binaries and freeimage.dll. I have compiled these on Windows and Mac using gcc and it works.

This requires FreeImage.dll to be stored in the same directory the executables are (Win32/Win64) or the FreeImage library to be linked to it (Linux/Unix/MacOS) somehow.

Two things to note about PNGs:
+Assumes PNG palette index #255 is transparent even if not marked so in Photoshop or Paint Shop Pro. If the image isn't indexed properly, the ART files will mess up. I should fix this somehow.
+Thus, make sure to use an act file for making PNGs that "knows" this, so what's transparent to you is transparent to the png2art.

Here is the tool description and syntax:

[ART2PNG]

This extracts the RAW art tiles to indexed PNGs. These PNGs can then be edited.

Syntax:

art2png numofartfiles palettefile inputdir outputdir

numofartfiles	-	total number of art files to process (usually 19 for DN3D Atomic)
palettefile		-	the file (only tested int current working directory) holding Duke 3D's PALETTE.DAT
inputdir		-	the directory where the art files are stored.
outputdir		-	the directory where the pngs will be stored, as well as animation data ini files.

for the directories, make sure they are created before populating or reading from them. mkdir can create directories from the command line on Windows

example syntax:

art2png 19 ./PALETTE.DAT ./tilesin ./pngout

[PNG2ART]

This populates RAW art tiles from indexed PNGs or full-color PNGs. PALETTE.DAT is used to aid conversion to 8-bit ART.

Valid PNGs include:

+8-bit Indexed PNGs
+24-bit full-colour PNGs
+24-bit full colour PNGs with an alpha Channel (aka 32-bit PNGs)

non-PNG formats are not supported and probably never will be.

8-bit indexed PNG images are not verified on a per-color basis, rather their pixels are copied into the art tile buffer.

Syntax:

png2art numofartfiles palettefile inputdir outputdir

numofartfiles	-	total number of art files to process (usually 19 for DN3D atomic)
inputdir		-	the directory where the pngs are stored, as well as animation data ini files.
outputdir		-	the directory where the art files will be created/overwritten.

for the directories, again, make sure they are created before populating/reading from them.

example syntax:

png2art 19 ./PALETTE.DAT ./pngin ./tilesout

Both assume that all files/pngs are going to need extracting/replaced/etc. It's recommended as this is alpha software to do a backup of any work.

These programs are released under the GPL license v3.

I am allowing people to beta-test this, offer constructive feedback. I am open to expanding these.

Look at changelog.txt and ideas.txt for more information.

CREDITS:

SanyaWaffles - Main coding
Mathieu Olivier - Original TGA-based tools, which inspired me to write these
Ken Silverman - Build Engine, documentation for ART format
eDuke32 team - for solving the riddle on how to convert PNGs to ART files
FreeImage - the mere 2MB (on Windows) library which made this all possible. Can be found here: http://freeimage.sourceforge.net/index.html


docs/changelog.txt
version 0.0.1 - Inital Version (12/13/2011)
	+ Initial Version and code
version 0.0.2 - (01/??/2012)
	+ Some minor fixes to code, documentation
version 0.1.0 - gcc compiled version (3/23/2012)
	+ This version contains many changes, especially in how it handles current working directories on *nix/MacOS, and Windows.
		It also changes how the file path delimiter is handled, conforming to research on the ANSI standard of C.
	+ No makefiles. To compile I recommend using a command line like:
		gcc -L/opt/usr/lib/ -I/opt/usr/include/ -lfreeimage -arch i386 -arch x86_64 art2png -o art2png
	+ This may need further testing on windows
	+ On mac, use universal macports version of the libfreeimage. Instructions on how to make global universal settings can be found on the eDuke32 compiling on OSX wiki article here:
		http://wiki.eduke32.com/wiki/Building_EDuke32_on_Mac_OS_X#MacPorts
	+ on *nix and Windows and MacOS, put the compiled binaries into your path.
		For Windows users, this might require putting freeimage.dll into your path as well.
version 0.1.2 - True Color PNG conversion
	+ png2art now reads a palette.dat to aid in color quantizing of PNG images with a greater BPP than 8-bits (24/32)

src/tests/trythis.c
#include <stdio.h>

#define		PATH_DELIMITER "/"

#ifdef _WIN32		// If we're on Win32/Win64

#include <direct.h>
#define GetCurrentDir _getcwd

#else				// If we're on *nix/Apple Mac OS X

#include <unistd.h>
#define GetCurrentDir getcwd

#endif

int main (int argc, char* argv[])
{
	char str[FILENAME_MAX];
	
	GetCurrentDir(str, sizeof(str));

	printf("The cwd is:\n\t\t%s%s\n", str, PATH_DELIMITER);
	fflush(stdout);

	return 0;
}

src/palgen.c
/* Copyright (C) 2012 SanyaWaffles
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */
 
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef		signed char		int8_t;
typedef		signed short	int16_t;
typedef		signed int		int32_t;
typedef	  unsigned char		uint8_t;
typedef	  unsigned short	uint16_t;
typedef	  unsigned int		uint32_t;

#ifndef __cplusplus
typedef enum {false, true} bool;
#endif

#define		PATH_DELIMITER "/"

#ifdef _WIN32		// If we're on Win32/Win64

#include <direct.h>
#define GetCurrentDir _getcwd

#else				// If we're on *nix/Apple Mac OS X

#include <unistd.h>
#define GetCurrentDir getcwd

#endif

#define TRANS_BYTES 65536
#define	PALETTEBYTES 768
#define	MAXPALOOKUPS 256

const char* PALETTE_FILENAMES[5] = {"main.act", "uwater.act",
						"slime.act", "title.act", "boss1.act"};

uint8_t		main_palette[PALETTEBYTES];
uint8_t		water_palette[PALETTEBYTES];
uint8_t		night_palette[PALETTEBYTES];
uint8_t		title_palette[PALETTEBYTES];
uint8_t		boss1_palette[PALETTEBYTES];
uint8_t		shadetables[MAXPALOOKUPS][PALETTEBYTES /3];
uint8_t		spr_tables[MAXPALOOKUPS][PALETTEBYTES / 3];
uint8_t		transdata[TRANS_BYTES];

uint32_t		shadetablenum;
uint32_t		spritepals;

char		cwd[FILENAME_MAX];
char		collectiondir[FILENAME_MAX];
char		outputdir[FILENAME_MAX];

static bool readPaletteDat(void);

static bool readLookupTable(void);

static bool dumpPalLookupScript(void);

static bool dumpACTPalettes(void);

// static void readPalLookupScript(void);

static bool readPaletteDat(void)
{
	uint32_t i, j;
	char palpath[FILENAME_MAX];
	FILE* pFile;
	
	sprintf(palpath, "%s%s%s%s%s", cwd, PATH_DELIMITER,
		collectiondir, PATH_DELIMITER, "PALETTE.DAT");
		
	// printf("%s\n", palpath);
	
	pFile = fopen(palpath, "rb");
	
	if (pFile == NULL)
	{
		printf("ERROR: PALETTE.DAT not found in \"%s\"\n", collectiondir);
		return false;
	}
	
	for (i = 0; i < PALETTEBYTES; i++)
	{
		main_palette[i] = fgetc(pFile);
	}
	
	shadetablenum = fgetc(pFile);
	
	if (shadetablenum > MAXPALOOKUPS)
	{
		printf("ERROR: SHADETABLENUM > 256!\n");
		return false;
	}
	
	for (i = 0; i <= shadetablenum; i++)
	{
		for (j = 0; j < (PALETTEBYTES / 3); j++)
		{
			shadetables[i][j] = fgetc(pFile);
		}
	}
	
	fclose(pFile);
	
	return true;
}

static bool readLookupTable(void)
{
	uint32_t i, j, id;
	char lookpath[FILENAME_MAX];
	FILE* luFile;
	
	sprintf(lookpath, "%s%s%s%s%s", cwd, PATH_DELIMITER,
		collectiondir, PATH_DELIMITER, "LOOKUP.DAT");
		
	// printf("%s\n", lookpath);
		
	luFile = fopen(lookpath, "rb");
	
	if (luFile == NULL)
	{
		printf("ERROR: LOOKUP.DAT not found\n");
		return false;
	}
	
	fread(&spritepals, sizeof(uint8_t), 1, luFile);
	
	if (spritepals > MAXPALOOKUPS && spritepals < 1)
	{
		printf("ERROR: SHADETABLENUM > 256!\n");
		return false;
	}
	
	for (i = 1; i <= spritepals; i++)
	{
		id = fgetc(luFile);
		printf("id parsed: %i\n", id);
		for (j = 0; j < (PALETTEBYTES / 3); j++)
		{
			spr_tables[id-1][j] = fgetc(luFile);
		}
	}
	
	for (i = 0; i < PALETTEBYTES; i++)
	{
		water_palette[i] = fgetc(luFile);
	}
	
	for (i = 0; i < PALETTEBYTES; i++)
	{
		night_palette[i] = fgetc(luFile);
	}
	
	for (i = 0; i < PALETTEBYTES; i++)
	{
		title_palette[i] = fgetc(luFile);
	}
	
	for (i = 0; i < PALETTEBYTES; i++)
	{
		boss1_palette[i] = fgetc(luFile);
	}
	
	fclose(luFile);
	
	return true;
}

static bool dumpPalLookupScript(void)
{
	uint32_t i, j, ors, ore, dss, dse;		
	uint8_t		difftables[spritepals][PALETTEBYTES / 3];
	bool printcomment;
	char scrpath[FILENAME_MAX];
	
	FILE* scrFile;
	
	sprintf(scrpath, "%s%s%s%s%s", cwd, PATH_DELIMITER, outputdir,
		PATH_DELIMITER, "pal_scr.txt");
		
	scrFile = fopen(scrpath, "wt");
	
	if (scrFile == NULL)
	{
		printf("ERROR: Cannot create pal_scr.txt\n");
		
		return false;
	}
	
	fprintf(scrFile, "EDUKE32_PALETTE_GENSCR\n");
	
	fprintf(scrFile, "[PALETTE.DAT]\n");
	
	fprintf(scrFile, "mainact = %s\n", PALETTE_FILENAMES[0]);
	
	fprintf(scrFile, "%s = %i\n\n", "shades", shadetablenum);
	
	fprintf(scrFile, "%s = %i\n\n", "updatetrans", 1);
	
	fprintf(scrFile, "[LOOKUP.DAT]\n");
	
	fprintf(scrFile, "%s = %i\n", "sprpals", spritepals);
	
	
	for (i = 0; i <= spritepals; i++)
	{
		fprintf(scrFile, "\n%s = %i\n\n", "palswap", i+1);
		if(!printcomment)
		{
			fprintf(scrFile, "; palgen can accept ranges in the form xxx:xxx=yyy:yyy\n");
			fprintf(scrFile, "; rather than individual ranges.\n\n");
			printcomment = true;
		}
		for (j = 0; j < (PALETTEBYTES / 3) ; j++)
		{
			if (spr_tables[i][j] != j)
			{
				fprintf(scrFile, "%i -> %i\n", j, spr_tables[i][j]);
			}
		}
	}
	
	
	fprintf(scrFile, "\nuwtract = %s\n", PALETTE_FILENAMES[1]);
	
	fprintf(scrFile, "nvs_act = %s\n", PALETTE_FILENAMES[2]);
	
	fprintf(scrFile, "ttl_act = %s\n", PALETTE_FILENAMES[3]);
	
	fprintf(scrFile, "bossact = %s\n\n", PALETTE_FILENAMES[4]);
	
	fclose(scrFile);
	
	return true;
}

static bool dumpACTPalettes(void)
{
	uint32_t i, j, temppal, succeed;
	char actpath[FILENAME_MAX];
	uint8_t *	currentPal;
	
	FILE* actFile;
	
	succeed = 0;
	
	for (i = 0; i < 5; i++)
	{
		sprintf(actpath, "%s%s%s%s%s", cwd, PATH_DELIMITER, outputdir,
			PATH_DELIMITER, PALETTE_FILENAMES[i]);
			
		actFile = fopen(actpath, "wb");
		
		if (actFile != NULL)
		{
		
			switch (i)
			{
				case 0:
					currentPal = main_palette;
					break;
				case 1:
					currentPal = water_palette;
					break;
				case 2:
					currentPal = night_palette;
					break;
				case 3:
					currentPal = title_palette;
					break;
				case 4:
					currentPal = boss1_palette;
					break;
				default:
					break;
				
			}
			
			if (currentPal != NULL)
			{
				for (j = 0; j < PALETTEBYTES; j++)
				{
					temppal = (uint8_t)currentPal[j] * 4;
					putc(temppal, actFile);
				}
			}
			else
			{
				printf("ERROR: invalid palette index %i\n", i);
			}
			
			fclose(actFile);
			
			succeed++;
		
		}
		else
		{
			printf("ERROR: Cannot create %s\n", PALETTE_FILENAMES[i]);
			continue;
		}
		
	}
	
	if (succeed == 5)
	{
		return true;
	}
	
	return false;
	
}

/* static void readPalLookupScript(void)
{
	uint8_t 
} */

int main(int argc, char* argv[])
{
	GetCurrentDir(cwd, sizeof(cwd));
	
	printf("\n"
		"palgen by Kraig Culp\n"
		"uses transpal code by Ken Silverman and JonoF and eDuke32\n\n");
		
	if (argc != 4 || !strcmp("--o", argv[1]) || !strcmp("--i", argv[1]))
	{
		printf("syntax: palgen -o|-i [palette.dat & lookup.dat dir] [pal_scr.txt dir]\n"
			"generate script ex: palgen -o ./grp ./palgen\n"
			"generate *.dat ex: palgen -i ./grp ./palgen\n"
			"-o = generate script || -i = generate .dat files\n\n");
			
		return EXIT_FAILURE;
	}
	
	sprintf(collectiondir, "%s", argv[2]);
		
	sprintf(outputdir, "%s", argv[3]);
	
	// printf("cwd: %s\n\n%s and %s\n", cwd, collectiondir, outputdir);
	
	if (!readPaletteDat())
	{
		printf("ERROR: readPaletteDat() failed with return false\n\n" );
		return EXIT_FAILURE;
	}
	
	if (!readLookupTable())
	{
		printf("ERROR: readLookupTable() failed with return false\n\n");
		return EXIT_FAILURE;
	}
	
	if (!dumpPalLookupScript())
	{
		printf("ERROR: dumpPalLookupScript() failed with return false\n\n");
		return EXIT_FAILURE;
	}
	
	if (!dumpACTPalettes())
	{
		printf("ERROR: dumpACTPalettes() failed with return false\n\n");
		return EXIT_FAILURE;
	}

	return EXIT_SUCCESS;
}

src/png2art.c
/* Copyright (C) 2011 SanyaWaffles
 * Based on code by Mathieu Olivier
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

#include <assert.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <FreeImage.h>

//
// Types and Constants
//

typedef		signed char		int8_t;
typedef		signed short	int16_t;
typedef		signed int		int32_t;
typedef	  unsigned char		uint8_t;
typedef	  unsigned short	uint16_t;
typedef	  unsigned int		uint32_t;

// Tile Struct
typedef struct {
	uint16_t sizex;
	uint16_t sizey;
	uint32_t animdata;
	uint32_t offset;
} tile_t;

// Line Type Struct
typedef enum {
	LINE_TYPE_UNKNOWN,
	LINE_TYPE_SECTION,
	LINE_TYPE_KEY_VALUE,
	LINE_TYPE_EOF
} linetype_e;


#define		PATH_DELIMITER "/"

#ifdef _WIN32		// If we're on Win32/Win64

#include <direct.h>
#define GetCurrentDir _getcwd

#else				// If we're on *nix/Apple Mac OS X

#include <unistd.h>
#define GetCurrentDir getcwd

#endif

#ifndef __cplusplus
typedef enum {false, true} bool;
#endif

#define MAX_NUMBER_OF_TILES 9216	// Maximum number of tiles
#define MAX_KEY_SIZE 128			// Key size for parsing ini file keys
#define MAX_VALUE_SIZE 128			// Value size for parsing ini file values
#define PALETTE_SIZE (256 * 3)		// Palette size (768)

// Global Variables

static uint32_t numtiles = 0;					// Number of tiles
static uint32_t currfilenum = 0;				// Current .ART index
static uint32_t tilestartnum = 0;				// current file starting number
static uint8_t artfilenum = 0;					// Not sure. Redundant?
static uint8_t maxartfiles = 0;					// Maximum art tile
static uint32_t tileendnum = 255;				// Current file ending number
static tile_t TilesList[MAX_NUMBER_OF_TILES];	// list of tiles

// Animation types for Adata###.ini parser
static const char* animtypes[4] = {"none", "oscillation", "forward", "backward"};

// Holds palette. Currently unused
static uint8_t palette[PALETTE_SIZE];

static RGBQUAD rgbpal[256];

// Stores input/output directory strings
static char palfilestr[FILENAME_MAX];
static char inputdir[FILENAME_MAX];
static char outputdir[FILENAME_MAX];

// Method prototypes
static bool createArtFile(const char* afname);

static linetype_e extractAnimDataLine(FILE * animdatafile, char* key, char* value);

static void getAnimData(void);

static bool LoadPalette(char *pfname);

static uint16_t GetLittleEndianUInt16(const uint8_t* buffer);

static void SetLittleEndianUInt16(uint16_t integer, uint8_t* buffer);

static void SetLittleEndianUInt32(uint32_t integer, uint8_t* buffer);

static bool parsePNGFile(uint32_t pngi, FILE* afile);

//
// IMPLEMENTATIONS
//


// CreateArtFile()
// Creates an art file. Pulls in PNGs
static bool createArtFile(const char* afname)
{
	// Variables
	FILE* artfile;
	uint8_t buffer[16 + MAX_NUMBER_OF_TILES * (2 + 2 + 4)];
	uint32_t i;
	uint32_t j;
	char png[FILENAME_MAX];

	artfile = fopen(afname, "wb");
	if (artfile == NULL)
	{
		printf("error: cannot create %d\n", artfilenum);
		return false;
	}

	SetLittleEndianUInt32(1, &buffer[0]);
	SetLittleEndianUInt32(tilestartnum + numtiles, &buffer[4]);
	SetLittleEndianUInt32(tilestartnum, &buffer[8]);
	SetLittleEndianUInt32(tilestartnum + numtiles - 1, &buffer[12]);

	fwrite(buffer, 1, 16 + numtiles * (2 + 2 + 4), artfile);

	for (i = 0; i < numtiles; i++)
	{
		parsePNGFile(tilestartnum + i, artfile);
	}

	getAnimData();

	if (fseek(artfile, 16, SEEK_SET) != 0)
	{
		printf("error: can't go to beginning of art file to write it's header\n");
		fclose(artfile);
		return false;
	}

	for (i = 0; i < numtiles; i++)
		SetLittleEndianUInt16(TilesList[tilestartnum + i].sizex, &buffer[i * 2]);
	fwrite(buffer, 1, numtiles * 2, artfile);

	for (i = 0; i < numtiles; i++)
		SetLittleEndianUInt16(TilesList[tilestartnum + i].sizey, &buffer[i * 2]);
	fwrite(buffer, 1, numtiles * 2, artfile);

	for (i = 0; i < numtiles; i++)
		SetLittleEndianUInt32(TilesList[tilestartnum + i].animdata, &buffer[i * 4]);
	fwrite(buffer, 1, numtiles * 4, artfile);

	fclose(artfile);

	return true;
}

// extractAnimDataLine()
// Extracts a line of data for reading
static linetype_e extractAnimDataLine(FILE* animdatafile, char* key, char* value)
{
	char line[128];
	char* charptr;
	uint32_t ki, vi;

	while (fgets(line, sizeof(line), animdatafile) != NULL)
	{
		charptr = line;
		while (*charptr == ' ')
			charptr++;

		switch(*charptr)
		{
		case '[':
			charptr++;
			while (*charptr == ' ')
				charptr++;

			ki = 0;
			while (*charptr != ']' && *charptr != '\0')
				key[ki++] = *charptr++;
			key[ki] = '\0';
			value[0] = '\0';

			if (*charptr == '\0')
				return LINE_TYPE_UNKNOWN;

			while (ki > 0 && key[ki - 1] == ' ')
				key[--ki] = '\0';

			return LINE_TYPE_SECTION;

		case '\n':
		case ';':
			break;

		default:
			ki = 0;
			while (*charptr != '=' && *charptr != '\0')
				key[ki++] = *charptr++;
			key[ki] = '\0';
			value[0] = '\0';

			if (*charptr == '\0')
				return LINE_TYPE_UNKNOWN;

			while (ki > 0 && key[ki - 1] == ' ')
				key[--ki] = '\0';

			charptr++;
			while (*charptr == ' ')
				charptr++;

			vi = 0;
			while (*charptr != '\n' && *charptr != '\0')
				value[vi++] = *charptr++;
			value[vi] = '\0';

			while(vi > 0 && value[vi - 1] == ' ')
				value[--vi] = '\0';

			return LINE_TYPE_KEY_VALUE;
		}
	}
	
	return LINE_TYPE_EOF;

}

// getAnimData()
static void getAnimData(void)
{
	FILE* adfile;
	linetype_e ltype;
	uint32_t currtilei, tilei1, tilei2;
	int32_t integer;
	char key[MAX_KEY_SIZE], value[MAX_VALUE_SIZE];
	char curradatafile[FILENAME_MAX];
	char* charptr;
	bool validsel;

	sprintf(curradatafile, "%s%sadata%03u.ini", inputdir, PATH_DELIMITER, artfilenum);
	adfile = fopen(curradatafile, "rt");
	if (adfile == NULL)
	{
		printf("warning: %s not found. No animation data added\n", curradatafile);
		return;
	}
	printf("parsing...\n");

	currtilei = tilestartnum;
	ltype = extractAnimDataLine(adfile, key, value);
	while (ltype != LINE_TYPE_EOF)
	{
		switch (ltype)
		{
		case LINE_TYPE_KEY_VALUE:
			if (strcmp(key, "AnimationType") == 0)
			{
				for (integer = 0; integer < 4; integer++)
					if (strcmp(animtypes[integer], value) == 0)
						break;

				if (integer == 4)
				{
					printf("Error: invalid animation type (%s)\n", value);
					break;
				}
				TilesList[currtilei].animdata &= 0xFFFFFF3F;
				TilesList[currtilei].animdata |= (integer & 0x03) << 6;
			}
			else if (strcmp(key, "AnimationSpeed") == 0)
			{
				integer = atoi(value);
				TilesList[currtilei].animdata &= 0xF0FFFFFF;
				TilesList[currtilei].animdata |= (integer & 0x0F) << 24;
			}
			else if (strcmp(key, "XCenterOffset") == 0)
			{
				integer = atoi(value);
				TilesList[currtilei].animdata &= 0xFFFF00FF;
				TilesList[currtilei].animdata |= (uint8_t)((int8_t)integer) << 8;
			}
			else if (strcmp(key, "YCenterOffset") == 0)
			{
				integer = atoi(value);
				TilesList[currtilei].animdata &= 0xFF00FFFF;
				TilesList[currtilei].animdata |= (uint8_t)((int8_t)integer) << 16;
			}
			else if (strcmp(key, "OtherFlags") == 0)
			{
				integer = atoi(value);
				TilesList[currtilei].animdata &= 0x0FFFFFFF;
				TilesList[currtilei].animdata |= (uint8_t)integer << 28;
			}
			else
				printf("error: unknown key %s\n", key);
			break;


		case LINE_TYPE_SECTION:
			do
			{
				validsel = false;

				charptr = strstr(key, "->");
				if (charptr != NULL)
				{
					if (sscanf(key, "tile%u.png -> tile%u.png", &tilei1, &tilei2) != 2)
						printf("Error: invalid selection name (%s)\n", key);
					else if (tilei1 < tilestartnum || tilei1 >= tilestartnum + numtiles)
						printf("error: invalid tile number (%u)\n", tilei1);
					else if (tilei2 < tilestartnum || tilei2 >= tilestartnum + numtiles)
						printf("Error: invalid tile number (%u)\n", tilei2);
					else if (tilei1 > tilei2)
						printf("Error: first tile number greater than second one (%u > %u)\n", tilei1, tilei2);
					else
					{
						currtilei = tilei1;
						TilesList[currtilei].animdata &= 0xFFFFFFC0;
						TilesList[currtilei].animdata |= (tilei2 - tilei1) & 0x3F;
						validsel = true;
					}
				}

				else
				{
					if (sscanf(key, "tile%u.png", &tilei1) != 1)
						printf("error: invalid section name (%s)\n", key);
					else if (tilei1 < tilestartnum || tilei1 >= tilestartnum + numtiles)
						printf("error: invalid tilenumber (%u)\n", tilei1);
					else
					{
						currtilei = tilei1;
						validsel = true;
					}
				}

				if (!validsel)
				{
					ltype = extractAnimDataLine(adfile, key, value);
					while (ltype != LINE_TYPE_SECTION)
					{
						if (ltype != LINE_TYPE_EOF)
						{
							fclose(adfile);
							return;
						}
						ltype = extractAnimDataLine(adfile, key, value);
					}
				}

			} while(!validsel) ;
			break;
		default:
			assert(ltype == LINE_TYPE_UNKNOWN);
			printf("    Warning: syntax error while parsing line %s\n", key);
			break;
			}
		ltype = extractAnimDataLine(adfile, key, value);
	}
	fclose(adfile);
}

static bool LoadPalette(char *pfname)
{
	// variables
	FILE* pfile;

	uint32_t i;
	uint8_t color;

	pfile = fopen(pfname, "rb");
	if (pfile == NULL)
	{
		printf("warning: cannot open palette.dat\n");
		return false;
	}

	// read palette
	if (fread(palette, 1, PALETTE_SIZE, pfile) != PALETTE_SIZE)
	{
		printf("warning: cannot read the whole palette from palette.dat file\n");
		fclose(pfile);
		return false;
	}

	for (i = 0; i < PALETTE_SIZE; i += 3)
	{
		rgbpal[i/3].rgbRed = palette[i] * 4;
		rgbpal[i/3].rgbGreen = palette[i + 1] * 4;
		rgbpal[i/3].rgbBlue = palette[i + 2] * 4;
	}
	
	if (rgbpal[255].rgbRed != 0 && rgbpal[255].rgbGreen != 255 && rgbpal[255].rgbBlue != 255)
	{
		rgbpal[255].rgbRed = 0;
		rgbpal[255].rgbGreen = 247;
		rgbpal[255].rgbBlue = 247;
	}
	
}

// GetLittleEndianUInt16()
// returns a bitshifted LittleEndian 16-bit int
static uint16_t GetLittleEndianUInt16(const uint8_t* buffer)
{
	return (uint16_t)(buffer[0] | (buffer[1] << 8));
}

// Main method
int main(int argc, char* argv[])
{
	char cwd[FILENAME_MAX];
	char path[FILENAME_MAX];	// Temp path
	int8_t tempnum;

	FreeImage_Initialise(0);	// We have to initialize FreeImage library before anything else.
	
	GetCurrentDir(cwd, sizeof(cwd));

	printf("\n"
		"png2art by Kraig Culp\n"
		"based on tga2art by Matthieu Oliver\n\n");

	if (argc != 5)
	{
		printf("syntax: png2art ## palette indir outdir\n"
			"ex: png2art 19 palette.dat pngs newart\n\n");
		FreeImage_DeInitialise();
		return EXIT_FAILURE;
	}

	tilestartnum = 0;
	tileendnum = 255;

	tempnum = atoi(argv[1]);
	
	// check if it's in range
	/* if (tempnum < 0)
		tempnum = 0;
	else if (tempnum > 255)
		tempnum = 255; */
	
	maxartfiles = tempnum;
	sprintf(palfilestr, "%s%s%s", cwd, PATH_DELIMITER, argv[2]);
	sprintf(inputdir, "%s%s%s", cwd, PATH_DELIMITER, argv[3]);
	sprintf(outputdir, "%s%s%s", cwd, PATH_DELIMITER, argv[4]);
	
	if (!LoadPalette(palfilestr))
	{
		FreeImage_DeInitialise();
		return EXIT_FAILURE;
	}

	numtiles = tileendnum - tilestartnum + 1;
	if (tilestartnum > tileendnum)
	{
		printf("error: the number of the first tile is greater than that of the last one!\n");
		FreeImage_DeInitialise();
		return EXIT_FAILURE;
	}

	for (artfilenum = 0; artfilenum <= maxartfiles; artfilenum++)
	{
		tilestartnum = (artfilenum * 256);
		tileendnum = tilestartnum + 256 - 1;

		numtiles = tileendnum - tilestartnum + 1;

		sprintf(path, "%s%sTILES%03u.art", outputdir, PATH_DELIMITER, artfilenum);
		if (!createArtFile(path))
		{
			FreeImage_DeInitialise();
			return EXIT_FAILURE;
		}
	}

	FreeImage_DeInitialise();
	return EXIT_SUCCESS;
}

static void SetLittleEndianUInt16(uint16_t integer, uint8_t* buffer)
{
	buffer[0] = (uint8_t)(integer & 255);
	buffer[1] = (uint8_t)(integer >> 8);
}

static void SetLittleEndianUInt32(uint32_t integer, uint8_t* buffer)
{
	buffer[0] = (uint8_t)(integer & 255);
	buffer[1] = (uint8_t)((integer >> 8) & 255);
	buffer[2] = (uint8_t)((integer >> 16) & 255);
	buffer[3] = (uint8_t)(integer >> 24);
}

// parsePNGFile()
// Takes in a PNG file and plugs the indexes into
// the proper art format.
static bool parsePNGFile(uint32_t pngi, FILE* afile)
{
	FILE* pngfile;
	char pngfilename[FILENAME_MAX];
	uint8_t* buffer;
	uint32_t xsize, ysize;
	int32_t xi, yi;
	FIBITMAP *pngas;
	FIBITMAP *pngaspal;
	FIBITMAP *pngasbg;
	FIBITMAP *pngas24;
	FIBITMAP *pngastemp;
	uint32_t m;
	uint32_t n;
	uint32_t _px;
	uint32_t* _pbx;
	
	pngi;

	TilesList[pngi].animdata = 0;
	TilesList[pngi].offset = ftell(afile);
	TilesList[pngi].sizex = 0;
	TilesList[pngi].sizey = 0;

	sprintf(pngfilename, "%s%stile%04u.png", inputdir, PATH_DELIMITER, pngi);
	
	pngastemp = FreeImage_Load(FIF_PNG, pngfilename, 0);

	if (pngastemp == NULL)
		return false;
		
	pngas = FreeImage_AllocateEx(FreeImage_GetWidth(pngastemp), 
		FreeImage_GetHeight(pngastemp), 8, &rgbpal[255], 0, rgbpal, 0, 0, 0);

	FreeImage_FlipVertical(pngastemp);
	
	if (FreeImage_GetBPP(pngastemp) != 8)
	{
		// printf("ssssnaaaaoooo\n");
		
		if (FreeImage_GetBPP(pngastemp) == 32)
		{
			pngasbg = FreeImage_Composite(pngastemp, FALSE, &rgbpal[255], NULL);
			// pngas24 = FreeImage_ConvertTo24Bits(pngasbg);
			
			
			pngaspal = FreeImage_ColorQuantizeEx(pngasbg, FIQ_NNQUANT, 256, 256, rgbpal);
			FreeImage_SetTransparentIndex(pngaspal, 255);
		}
		else
		{
			pngaspal = FreeImage_ColorQuantizeEx(pngastemp, FIQ_NNQUANT, 256, 256, rgbpal);
			FreeImage_SetTransparentIndex(pngaspal, 255);
		}
		
		if (pngaspal == NULL)
		{
			printf("error: tile%04u.png is an invalid 8/24/32bit image\n", pngi);
			return false;
		}	
			
		pngas = pngaspal;
		
	}
	else
	{
		pngas = pngastemp;
	}
	
	xsize = FreeImage_GetWidth(pngas);
	ysize = FreeImage_GetHeight(pngas);

	TilesList[pngi].sizex = xsize;
	TilesList[pngi].sizey = ysize;

	buffer = malloc(xsize * ysize);
	
	if (buffer == NULL)
	{
		printf("error: not enough memory to read image\n");
		return false;
	}

	// This is where the magic happens
	for (xi = 0; xi  < TilesList[pngi].sizex; xi++)
	{
		for (yi = 0; yi < TilesList[pngi].sizey; yi++)
		{
			FreeImage_GetPixelIndex(pngas, xi, yi, &buffer[xi * TilesList[pngi].sizey + yi]);
			fwrite(&buffer[xi * TilesList[pngi].sizey + yi], 1, 1, afile);
		}
	}
	
	free(buffer);	// Gotta free memory explicitly

	return true;
}